File: src/Api/Services/EngagementService.cs
using System.Net.Http.Json;
using System.Text.Json;
using Microsoft.EntityFrameworkCore;
using PostGenerator.Api.Data;
using PostGenerator.Api.Models;
using PostGenerator.Core;

namespace PostGenerator.Api.Services;

public class EngagementService : IEngagementService
{
    private readonly AppDbContext _db;
    private readonly IHttpClientFactory _httpClientFactory;
    private readonly ILogger<EngagementService> _logger;

    public EngagementService(AppDbContext db, IHttpClientFactory httpClientFactory, ILogger<EngagementService> logger)
    {
        _db = db;
        _httpClientFactory = httpClientFactory;
        _logger = logger;
    }

    public async Task<PostDto?> RefreshEngagementAsync(int userId, int postId, CancellationToken cancellationToken = default)
    {
        var post = await _db.Posts.FirstOrDefaultAsync(p => p.Id == postId && p.UserId == userId, cancellationToken);
        if (post == null) return null;
        if (string.IsNullOrEmpty(post.ExternalPostId))
        {
            _logger.LogDebug("Post {PostId} has no ExternalPostId, cannot refresh engagement", postId);
            return null;
        }

        var cred = await _db.UserSocialCredentials
            .AsNoTracking()
            .Where(c => c.UserId == userId && c.Platform == post.Platform.ToString())
            .Select(c => c.CredentialJson)
            .FirstOrDefaultAsync(cancellationToken);
        if (string.IsNullOrEmpty(cred))
        {
            _logger.LogDebug("No credentials for user {UserId} platform {Platform}", userId, post.Platform);
            return null;
        }
        Dictionary<string, string>? credentials;
        try
        {
            credentials = JsonSerializer.Deserialize<Dictionary<string, string>>(cred);
        }
        catch
        {
            return null;
        }
        if (credentials == null || credentials.Count == 0) return null;

        int? views = null, likes = null, comments = null;
        try
        {
            switch (post.Platform)
            {
                case PostPlatform.Facebook:
                    (views, likes, comments) = await FetchFacebookEngagementAsync(post.ExternalPostId, credentials, cancellationToken);
                    break;
                case PostPlatform.Instagram:
                    (views, likes, comments) = await FetchInstagramEngagementAsync(post.ExternalPostId, credentials, cancellationToken);
                    break;
                case PostPlatform.Bluesky:
                    (views, likes, comments) = await FetchBlueskyEngagementAsync(post.ExternalPostId, credentials, cancellationToken);
                    break;
                case PostPlatform.LinkedIn:
                case PostPlatform.TikTok:
                case PostPlatform.Skool:
                    _logger.LogDebug("Engagement refresh not supported for platform {Platform}", post.Platform);
                    return null;
                default:
                    return null;
            }
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Engagement fetch failed for post {PostId} platform {Platform}", postId, post.Platform);
            return null;
        }

        post.ViewsCount = views ?? post.ViewsCount;
        post.LikesCount = likes ?? post.LikesCount;
        post.CommentsCount = comments ?? post.CommentsCount;
        post.LastEngagementFetchedAt = DateTime.UtcNow;
        post.UpdatedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync(cancellationToken);
        return PostMappings.ToDto(post);
    }

    private async Task<(int? Views, int? Likes, int? Comments)> FetchFacebookEngagementAsync(string postId, IReadOnlyDictionary<string, string> credentials, CancellationToken ct)
    {
        var token = credentials.TryGetValue("PageAccessToken", out var t) ? t : null;
        if (string.IsNullOrEmpty(token)) return (null, null, null);

        var client = _httpClientFactory.CreateClient();
        var url = $"https://graph.facebook.com/v21.0/{Uri.EscapeDataString(postId)}?fields=likes.summary(true),comments.summary(true)&access_token={Uri.EscapeDataString(token)}";
        using var resp = await client.GetAsync(url, ct);
        resp.EnsureSuccessStatusCode();
        var doc = await resp.Content.ReadFromJsonAsync<JsonElement>(cancellationToken: ct);
        var likes = doc.TryGetProperty("likes", out var likesEl) && likesEl.TryGetProperty("summary", out var likesSum) && likesSum.TryGetProperty("total_count", out var lc) ? lc.GetInt32() : (int?)null;
        var comments = doc.TryGetProperty("comments", out var commentsEl) && commentsEl.TryGetProperty("summary", out var commentsSum) && commentsSum.TryGetProperty("total_count", out var cc) ? cc.GetInt32() : (int?)null;

        int? views = null;
        var insightsUrl = $"https://graph.facebook.com/v21.0/{Uri.EscapeDataString(postId)}/insights?metric=post_impressions&access_token={Uri.EscapeDataString(token)}";
        using var insightsResp = await client.GetAsync(insightsUrl, ct);
        if (insightsResp.IsSuccessStatusCode)
        {
            var insights = await insightsResp.Content.ReadFromJsonAsync<JsonElement>(cancellationToken: ct);
            if (insights.TryGetProperty("data", out var data) && data.GetArrayLength() > 0)
            {
                var first = data[0];
                if (first.TryGetProperty("values", out var values) && values.GetArrayLength() > 0 && values[0].TryGetProperty("value", out var v))
                    views = v.GetInt32();
            }
        }
        return (views, likes, comments);
    }

    private async Task<(int? Views, int? Likes, int? Comments)> FetchInstagramEngagementAsync(string mediaId, IReadOnlyDictionary<string, string> credentials, CancellationToken ct)
    {
        var token = credentials.TryGetValue("AccessToken", out var at) ? at : null;
        if (string.IsNullOrEmpty(token)) return (null, null, null);

        var client = _httpClientFactory.CreateClient();
        client.DefaultRequestHeaders.Add("Authorization", "Bearer " + token);
        var url = $"https://graph.facebook.com/v21.0/{Uri.EscapeDataString(mediaId)}/insights?metric=engagement,impressions,reach";
        using var resp = await client.GetAsync(url, ct);
        resp.EnsureSuccessStatusCode();
        var doc = await resp.Content.ReadFromJsonAsync<JsonElement>(cancellationToken: ct);
        int? views = null, likes = null, comments = null;
        if (doc.TryGetProperty("data", out var data))
        {
            foreach (var item in data.EnumerateArray())
            {
                if (!item.TryGetProperty("name", out var nameEl) || !item.TryGetProperty("values", out var values) || values.GetArrayLength() == 0) continue;
                var value = values[0].TryGetProperty("value", out var v) ? v.GetInt32() : 0;
                switch (nameEl.GetString())
                {
                    case "impressions": views = value; break;
                    case "engagement": likes = value; break;
                    case "reach": if (!views.HasValue) views = value; break;
                }
            }
        }
        return (views, likes, comments);
    }

    private async Task<(int? Views, int? Likes, int? Comments)> FetchBlueskyEngagementAsync(string postUri, IReadOnlyDictionary<string, string> credentials, CancellationToken ct)
    {
        var handle = credentials.TryGetValue("Handle", out var h) ? h : null;
        var appPassword = credentials.TryGetValue("AppPassword", out var p) ? p : null;
        var pdsUrl = (credentials.TryGetValue("PdsUrl", out var u) ? u : null)?.TrimEnd('/') ?? "https://bsky.social";
        if (string.IsNullOrEmpty(handle) || string.IsNullOrEmpty(appPassword)) return (null, null, null);

        var client = _httpClientFactory.CreateClient();
        var sessionUrl = $"{pdsUrl}/xrpc/com.atproto.server.createSession";
        var sessionBody = new { identifier = handle, password = appPassword };
        using var sessionResp = await client.PostAsJsonAsync(sessionUrl, sessionBody, ct);
        sessionResp.EnsureSuccessStatusCode();
        var sessionJson = await sessionResp.Content.ReadFromJsonAsync<JsonElement>(cancellationToken: ct);
        var accessJwt = sessionJson.GetProperty("accessJwt").GetString();
        if (string.IsNullOrEmpty(accessJwt)) return (null, null, null);

        const int maxPages = 10;
        int likeCount = 0;
        string? cursor = null;
        var pageCount = 0;
        do
        {
            if (++pageCount > maxPages) break;
            var likesUrl = $"{pdsUrl}/xrpc/app.bsky.feed.getLikes?uri={Uri.EscapeDataString(postUri)}&limit=100";
            if (!string.IsNullOrEmpty(cursor)) likesUrl += "&cursor=" + Uri.EscapeDataString(cursor);
            using var request = new HttpRequestMessage(HttpMethod.Get, likesUrl);
            request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", accessJwt);
            using var likesResp = await client.SendAsync(request, ct);
            likesResp.EnsureSuccessStatusCode();
            var likesJson = await likesResp.Content.ReadFromJsonAsync<JsonElement>(cancellationToken: ct);
            if (likesJson.TryGetProperty("likes", out var likes))
                likeCount += likes.GetArrayLength();
            cursor = likesJson.TryGetProperty("cursor", out var c) ? c.GetString() : null;
        } while (!string.IsNullOrEmpty(cursor));

        return (null, likeCount, null);
    }
}

File: src/lib/services/engagement-service.ts
import { db } from "@/db/drizzle";
import { posts, userSocialCredentials } from "@/db/schema";
import { eq, and } from "drizzle-orm";
import type { PostDto } from "./post-service";

function toDto(p: {
  id: number;
  userId: number;
  topicSummary: string | null;
  platform: string;
  status: string;
  scheduledAt: Date | null;
  publishedAt: Date | null;
  externalPostId: string | null;
  viewsCount: number | null;
  likesCount: number | null;
  commentsCount: number | null;
  lastEngagementFetchedAt: Date | null;
  content: string;
  script: string | null;
  imageUrl: string | null;
  metadataJson: string | null;
  tone: string | null;
  length: string | null;
  createdAt: Date;
  updatedAt: Date;
}): PostDto {
  return {
    id: p.id,
    userId: p.userId,
    topicSummary: p.topicSummary ?? "",
    platform: p.platform,
    status: p.status,
    scheduledAt: p.scheduledAt?.toISOString() ?? null,
    publishedAt: p.publishedAt?.toISOString() ?? null,
    externalPostId: p.externalPostId,
    viewsCount: p.viewsCount,
    likesCount: p.likesCount,
    commentsCount: p.commentsCount,
    lastEngagementFetchedAt: p.lastEngagementFetchedAt?.toISOString() ?? null,
    content: p.content,
    script: p.script,
    imageUrl: p.imageUrl,
    metadataJson: p.metadataJson,
    tone: p.tone,
    length: p.length,
    createdAt: p.createdAt.toISOString(),
    updatedAt: p.updatedAt.toISOString(),
  };
}

export async function refreshEngagement(
  userId: number,
  postId: number
): Promise<PostDto | null> {
  const [post] = await db
    .select()
    .from(posts)
    .where(and(eq(posts.id, postId), eq(posts.userId, userId)))
    .limit(1);

  if (!post || !post.externalPostId) return null;

  const [credRow] = await db
    .select()
    .from(userSocialCredentials)
    .where(
      and(
        eq(userSocialCredentials.userId, userId),
        eq(userSocialCredentials.platform, post.platform)
      )
    )
    .limit(1);

  if (!credRow?.credentialJson) return null;

  let credentials: Record<string, string>;
  try {
    credentials = JSON.parse(credRow.credentialJson) as Record<string, string>;
  } catch {
    return null;
  }
  if (Object.keys(credentials).length === 0) return null;

  let views: number | null = null;
  let likes: number | null = null;
  let comments: number | null = null;

  try {
    if (post.platform === "Facebook") {
      const r = await fetchFacebookEngagement(post.externalPostId, credentials);
      views = r.views;
      likes = r.likes;
      comments = r.comments;
    } else if (post.platform === "Instagram") {
      const r = await fetchInstagramEngagement(post.externalPostId, credentials);
      views = r.views;
      likes = r.likes;
    } else if (post.platform === "Bluesky") {
      likes = await fetchBlueskyEngagement(userId, post.externalPostId, credentials);
    }
  } catch {
    return null;
  }

  const now = new Date();
  await db
    .update(posts)
    .set({
      viewsCount: views ?? post.viewsCount,
      likesCount: likes ?? post.likesCount,
      commentsCount: comments ?? post.commentsCount,
      lastEngagementFetchedAt: now,
      updatedAt: now,
    })
    .where(eq(posts.id, postId));

  const [updated] = await db
    .select()
    .from(posts)
    .where(eq(posts.id, postId))
    .limit(1);

  return updated ? toDto(updated) : null;
}

async function fetchFacebookEngagement(
  postId: string,
  credentials: Record<string, string>
): Promise<{ views: number | null; likes: number | null; comments: number | null }> {
  const token = credentials.PageAccessToken;
  if (!token) return { views: null, likes: null, comments: null };

  const url = `https://graph.facebook.com/v21.0/${encodeURIComponent(postId)}?fields=likes.summary(true),comments.summary(true)&access_token=${encodeURIComponent(token)}`;
  const res = await fetch(url);
  if (!res.ok) throw new Error(`Facebook API ${res.status}`);

  const doc = (await res.json()) as Record<string, unknown>;
  const likesEl = (doc.likes as Record<string, unknown>)?.summary as Record<string, unknown> | undefined;
  const commentsEl = (doc.comments as Record<string, unknown>)?.summary as Record<string, unknown> | undefined;
  const likes = typeof likesEl?.total_count === "number" ? likesEl.total_count : null;
  const comments = typeof commentsEl?.total_count === "number" ? commentsEl.total_count : null;

  const insightsUrl = `https://graph.facebook.com/v21.0/${encodeURIComponent(postId)}/insights?metric=post_impressions&access_token=${encodeURIComponent(token)}`;
  const insightsRes = await fetch(insightsUrl);
  let views: number | null = null;
  if (insightsRes.ok) {
    const insights = (await insightsRes.json()) as Record<string, unknown>;
    const data = insights.data as Array<{ values?: Array<{ value?: number }> }> | undefined;
    if (data?.[0]?.values?.[0]?.value != null) {
      views = data[0].values[0].value;
    }
  }

  return { views, likes, comments };
}

async function fetchInstagramEngagement(
  mediaId: string,
  credentials: Record<string, string>
): Promise<{ views: number | null; likes: number | null }> {
  const token = credentials.AccessToken;
  if (!token) return { views: null, likes: null };

  const url = `https://graph.facebook.com/v21.0/${encodeURIComponent(mediaId)}/insights?metric=engagement,impressions,reach`;
  const res = await fetch(url, {
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!res.ok) throw new Error(`Instagram API ${res.status}`);

  const doc = (await res.json()) as { data?: Array<{ name?: string; values?: Array<{ value?: number }> }> };
  let views: number | null = null;
  let likes: number | null = null;
  for (const item of doc.data ?? []) {
    const value = item.values?.[0]?.value ?? 0;
    if (item.name === "impressions") views = value;
    else if (item.name === "engagement") likes = value;
    else if (item.name === "reach" && views == null) views = value;
  }
  return { views, likes };
}

const blueskyCache = new Map<string, { jwt: string; expires: number }>();
const CACHE_TTL_MS = 5 * 60 * 1000;

async function fetchBlueskyEngagement(
  userId: number,
  postUri: string,
  credentials: Record<string, string>
): Promise<number | null> {
  const handle = credentials.Handle;
  const appPassword = credentials.AppPassword;
  const pdsUrl = (credentials.PdsUrl ?? "https://bsky.social").replace(/\/$/, "");

  if (!handle || !appPassword) return null;

  const cacheKey = `${userId}:Bluesky`;
  let entry = blueskyCache.get(cacheKey);
  if (!entry || entry.expires < Date.now()) {
    const sessionRes = await fetch(`${pdsUrl}/xrpc/com.atproto.server.createSession`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ identifier: handle, password: appPassword }),
    });
    if (!sessionRes.ok) throw new Error(`Bluesky session ${sessionRes.status}`);

    const session = (await sessionRes.json()) as { accessJwt?: string };
    const jwt = session.accessJwt;
    if (!jwt) return null;

    entry = { jwt, expires: Date.now() + CACHE_TTL_MS };
    blueskyCache.set(cacheKey, entry);
  }

  let likeCount = 0;
  let cursor: string | undefined;
  for (let page = 0; page < 10; page++) {
    let url = `${pdsUrl}/xrpc/app.bsky.feed.getLikes?uri=${encodeURIComponent(postUri)}&limit=100`;
    if (cursor) url += `&cursor=${encodeURIComponent(cursor)}`;

    const likesRes = await fetch(url, {
      headers: { Authorization: `Bearer ${entry.jwt}` },
    });
    if (!likesRes.ok) throw new Error(`Bluesky likes ${likesRes.status}`);

    const likesJson = (await likesRes.json()) as { likes?: unknown[]; cursor?: string };
    if (Array.isArray(likesJson.likes)) likeCount += likesJson.likes.length;
    cursor = likesJson.cursor;
    if (!cursor) break;
  }

  return likeCount;
}
